VirtoCommerce.Platform.Core/Common/IQueryableExtensions.cs

```diff
@@ -69,39 +69,40 @@ public static IOrderedQueryable<T> ThenBySortInfos<T>(this IOrderedQueryable<T>
 
         public static IOrderedQueryable<T> ApplyOrder<T>(IQueryable<T> source, string property, string methodName)
 		{
-			if (property == null)
-				throw new ArgumentNullException("property");
+            if (property == null)
+                throw new ArgumentNullException("property");
 
-			string[] props = property.Split('.');
-			Type type = typeof(T);
-			ParameterExpression arg = Expression.Parameter(type, "x");
-			Expression expr = arg;
+            string[] props = property.Split('.');
+            var effectiveType = source.FirstOrDefault().GetType();
```

Comment:
FirstOrDefault() will lead to redundant and unexpected request to the database.

Reply:
Really, it is non-acceptable solution. Cannot find a way to get specific element runtime type besides getting element. [IQueryable.ElementType](https://docs.microsoft.com/en-US/dotnet/api/system.linq.iqueryable.elementtype?view=netframework-4.7.2) does not help.  [System.Linq.Dynamic](https://github.com/kahanu/System.Linq.Dynamic) cannot handle such situations too.
So we should admit that we cannot implement sorting using derived class new properties in this extension without immediate execution.
The only viable option I see is passing type as propertyName, in string. 
So need to decide if additional extension like .OrderBy(string propertyName, string elementTypeName) could be useful for us.

Reply:
Decided to use type registered in AbstractTypeFactory for base type.

Reply:
@artem-dudarev Could you check new implementation?